File: src/main/java/eu/happybe/openapi/ranks/RankDatabase.java
package eu.happybe.openapi.ranks;

import cn.nukkit.Player;
import eu.happybe.openapi.OpenAPI;
import eu.happybe.openapi.mysql.QueryQueue;
import eu.happybe.openapi.mysql.query.UpdateRowQuery;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RankDatabase {

    private static final Map<String, Rank> ranks = new HashMap<>();

    public static void init() {
        List<Rank> ranks = Arrays.asList(
                // Staff
                new Rank("Owner", "§6§lOWNER", new String[]{"happybe.operator", "nukkit.command.gamemode", "nukkit.command.teleport", "nukkit.command.kick"}),
                new Rank("Developer", "§6§lDEVELOPER", new String[] {"happybe.operator"}),
                new Rank("Admin", "§6§lADMIN", new String[] {"happybe.operator", "nukkit.command.teleport", "nukkit.command.kick"}),
                new Rank("Mod", "§e§lMOD", new String[] {"happybe.moderator", "nukkit.command.teleport", "nukkit.command.kick"}),
                new Rank("Helper", "§e§lHELPER", new String[] {"happybe.helper", "nukkit.command.kick"}),
                new Rank("Builder", "§e§lBUILDER", new String[] {"happybe.builder"}),
                // Buyable ranks
                new Rank("Bedrock", "§9§lBEDROCK", new String[] {"happybe.bedrock"}),
                new Rank("MVP", "§3§lMVP§B+", new String[] {"happybe.mvp"}),
                new Rank("VIP", "§b§lVIP", new String[] {"happybe.vip"}),
                // Gettable ranks
                new Rank("YouTube", "§c§lYOUTUBE", new String[] {"happybe.bedrock"}),
                new Rank("Voter", "§b§lVOTER", new String[] {"happybe.voter"}),
                // Guest
                new Rank("Guest", "", new String[] {}, false)
        );

        for(Rank rank : ranks) {
            RankDatabase.ranks.put(rank.getRankName().toLowerCase(), rank);
        }
    }

    public static void savePlayerRank(Player player, String rank) {
        RankDatabase.savePlayerRank(player, rank, false);
    }

    public static void savePlayerRank(Player player, String rank, boolean saveToDatabase) {
        Rank targetRank = RankDatabase.getRankByName(rank);
        if(targetRank == null) {
            player.kick("Invalid rank (" + rank + ")");
            OpenAPI.getInstance().getLogger().error("Invalid rank received from database (" + rank + ") for player " + player.getName() + "!");
            return;
        }

        player.namedTag.putString("Rank", targetRank.getRankName());

        player.recalculatePermissions();
        for(String permission : targetRank.getPermissions()) {
            player.addAttachment(OpenAPI.getInstance(), permission, true);
        }

        if(saveToDatabase) {
            QueryQueue.submitQuery(new UpdateRowQuery(new HashMap<String, Object>() {{ this.put("Rank", targetRank.getRankName()); }}, "Name", player.getName()));
        }
    }

    public static Rank getPlayerRank(Player player) {
        return RankDatabase.getRankByName(player.namedTag.getString("Rank"));
    }

    public static Rank getRankByName(String name) {
        return ranks.get(name.toLowerCase());
    }
}

File: src/main/java/eu/happybe/openapi/ranks/RankDatabase.kt
package eu.happybe.openapi.ranks

import cn.nukkit.Player
import eu.happybe.openapi.OpenAPI
import eu.happybe.openapi.mysql.QueryQueue
import eu.happybe.openapi.mysql.query.UpdateRowQuery
import java.util.*

object RankDatabase {
    private val ranks: MutableMap<String, Rank> = HashMap()
    fun init() {
        val ranks = Arrays.asList( // Staff
                Rank("Owner", "§6§l", arrayOf("happybe.operator", "nukkit.command.gamemode", "nukkit.command.teleport", "nukkit.command.kick")),
                Rank("Developer", "§6§l", arrayOf("happybe.operator")),
                Rank("Admin", "§6§l", arrayOf("happybe.operator", "nukkit.command.teleport", "nukkit.command.kick")),
                Rank("Mod", "§e§l", arrayOf("happybe.moderator", "nukkit.command.teleport", "nukkit.command.kick")),
                Rank("Helper", "§e§l", arrayOf("happybe.helper", "nukkit.command.kick")),
                Rank("Builder", "§e§l", arrayOf("happybe.builder")),  // Buyable ranks
                Rank("Bedrock", "§9§l", arrayOf("happybe.bedrock")),
                Rank("MVP", "§3§l", arrayOf("happybe.mvp")),
                Rank("VIP", "§3§l", arrayOf("happybe.vip")),  // Gettable ranks
                Rank("YouTube", "§c§l", arrayOf("happybe.bedrock")),
                Rank("Voter", "§b§l", arrayOf("happybe.voter")),  // Guest
                Rank("Guest", "§b§l", arrayOf(), false)
        )
        for (rank in ranks) {
            RankDatabase.ranks[rank.rankName.toLowerCase()] = rank
        }
    }

    fun savePlayerRank(player: Player, rank: String) {
        savePlayerRank(player, rank, false)
    }

    fun savePlayerRank(player: Player, rank: String, saveToDatabase: Boolean) {
        val targetRank = getRankByName(rank)
        if (targetRank == null) {
            player.kick("Invalid rank ($rank)")
            OpenAPI.getInstance().logger.error("Invalid rank received from database (" + rank + ") for player " + player.name + "!")
            return
        }
        player.namedTag.putString("Rank", targetRank.rankName)
        player.recalculatePermissions()
        for (permission in targetRank.permissions) {
            player.addAttachment(OpenAPI.getInstance(), permission, true)
        }
        if (saveToDatabase) {
            QueryQueue.submitQuery(UpdateRowQuery(object : HashMap<String?, Any?>() {
                init {
                    this["Rank"] = targetRank.rankName
                }
            }, "Name", player.name))
        }
    }

    fun getPlayerRank(player: Player): Rank? {
        return getRankByName(player.namedTag.getString("Rank"))
    }

    fun getRankByName(name: String): Rank? {
        return ranks[name.toLowerCase()]
    }
}